### ISSUE TITLE
Add HTTP 103 Early Hints support

### ISSUE THREAD

**Issue by otbutz** [caddyserver/caddy#4860]
[RFC 8297](https://httpwg.org/specs/rfc8297.html) is only a draft, but is now supported by Chrome 103:

https://caniuse.com/mdn-http_status_103

and Mozilla also intends to support it:

https://bugzilla.mozilla.org/show_bug.cgi?id=1407355

Apache already implements it: https://httpd.apache.org/docs/2.4/howto/http2.html#earlyhints

It would be nice if caddy would add (experimental) support for it. Following the RFC recommendations, this should only be added for HTTP/2 and HTTP/3:

> Therefore, a server might refrain from sending 103 (Early Hints) responses over HTTP/1.1 unless the client is known to handle informational responses correctly.

**Comment by otbutz** [comment]
The very first step would be to make sure that caddy processes HTTP 103 properly while acting as a reverse proxy. That may already be the case.

**Comment by francislavoie** [comment]
@dunglas has been working on adding support for early hints to the Go stdlib, I'm sure he'll follow up soon to make sure Caddy supports it sufficiently.

**Comment by dunglas** [comment]
Indeed it's on my todo list. We'll have to wait for Go 1.19 first.

**Comment by bt90** [comment]
@dunglas does this also add support for HTTP3 based connections?

**Comment by dunglas** [comment]
@bt90 yes! I also patched the HTTP/3 library: https://github.com/lucas-clemente/quic-go/pull/3047

**Comment by pc-erin** [comment]
Is there any documentation on how use this feature now that it's implemented?

If we do `respond 103` does that send hints but still maintain the connection for the main response?

**Comment by mholt** [comment]
@pc-erin HTTP 103 is for applications to emit. The only "application" Caddy really offers (at least, built-in) is the static file server; I have been considering building a feature that will parse HTML and send 103 responses accordingly (like for CSS, JS, and image resources), but maybe at a future date.

For now, this relays 103s from proxy backends down to the client.

What's your use case?

**Comment by bt90** [comment]
Using it in a reverse proxy scenario might also be neat.  Add a matcher for the main page and tell the clients to also load files X, Y and Z.

**Comment by mholt** [comment]
Proposal implemented in #5006 -- @pc-erin and @bt90 you may want to try it out!

**Comment by xfalcox** [comment]
Hey @mholt,

When using 103 hints in a more conventional way, with caddy being a reverse proxy to an application that sends those responses, one useful feature is caching the 103 responses from previous requests and replying with those learned cached behaviors in newer requests. At the moment, AFAIK only Cloudflare implements this, but it can be very useful for Caddy to have something like this.

Example:

1. User requests `/index`
2. Caddy receives the request and proxies it to underlying app
3. Underlying app runs some fast logic and responds with 103. **This takes 100ms**
4. Underlying finishes building the full response. **This takes 300ms**

If Caddy "learns" that requests to `/index` get a certain 103 response, it could reply with that ASAP on step 2, making this response happen 100ms earlier, which will result in a faster browsing experience if the resources in 103 are necessary for website interactivity.

Curious about your opinions on this @mholt.

**Comment by mholt** [comment]
@xfalcox We could do that... would require a new directive to enable it, I think. Probably best for someone to implement it as a separate plugin for now and see how popular and useful it is!
